### PR TITLE
Queue system to spawn as larva

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

If there is no burrowed larva when you try to respawn as a xenos, you'll get placed in a queue and you'll get priority when a larva spot is added

## Why It's Good For The Game

More fair than "the first to click get the larva"

## Changelog
:cl:
qol: Queue system to respawn as a xeno
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 1]
```diff
@@ -1168,3 +1160,27 @@ to_chat will check for valid clients itself already so no need to double check f
...
+	var/datum/job/xeno_job = SSjob.GetJobType(/datum/job/xenomorph)
+	var/stored_larva = xeno_job.total_positions - xeno_job.current_positions
+	while(stored_larva >= 1 && candidate.len)
+		var/mob/next_in_line = candidate[1]
+		candidate.Cut(1,2)
+		xeno_job.occupy_job_positions(1)
+		stored_larva--
+		if(!attempt_to_spawn_larva(next_in_line, TRUE))
+			xeno_job.free_job_positions(1)
+			INVOKE_ASYNC(src, .proc/give_larva_next_in_queue)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm]: If you are asking why async, i don't know

File: code/modules/mob/dead/observer/observer.dm [hunk 2]
```diff
@@ -49,6 +49,8 @@ GLOBAL_VAR_INIT(observer_default_invisibility, INVISIBILITY_OBSERVER)
 	var/ghost_orderhud = FALSE
 	var/ghost_vision = TRUE
 	var/ghost_orbit = GHOST_ORBIT_CIRCLE
+	///Position in the larva queue
+	var/larva_position
```
Comment (TiviPlus) [inline code/modules/mob/dead/observer/observer.dm:53]: Is this a bool? an integer? Whats the default?

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 3]
```diff
@@ -1168,3 +1159,39 @@ to_chat will check for valid clients itself already so no need to double check f
...
+///Add a mob to the candidate queue, the first mobs of the queue will have priority on new larva spots
+/datum/hive_status/normal/proc/add_to_larva_candidate_queue(mob/dead/observer/observer)
+	var/datum/job/xeno_job = SSjob.GetJobType(/datum/job/xenomorph)
+	var/stored_larva = xeno_job.total_positions - xeno_job.current_positions
+	if(stored_larva)
+		attempt_to_spawn_larva(observer)
+		return
+	if(candidate.Find(observer))
+		to_chat(observer, "<span class='warning'>You are already in queue!</span>")
+	candidate += observer
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:1172]: Lacking return

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 4]
```diff
@@ -1168,3 +1159,39 @@ to_chat will check for valid clients itself already so no need to double check f
...
+	observer.larva_position = candidate.len
+	to_chat(observer, "<span class='warning'>There are no burrowed larvas. You are in position [candidate.len] to become a xeno</span>")
+
+///Propose larvas until their is no more candidates, or no more burrowed
+/datum/hive_status/normal/proc/give_larva_to_next_in_queue()
+	if(!candidate.len)
+		return
+	var/datum/job/xeno_job = SSjob.GetJobType(/datum/job/xenomorph)
+	var/stored_larva = xeno_job.total_positions - xeno_job.current_positions
+	while(stored_larva >= 1 && candidate.len)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:1182]:
```suggestion
	while(stored_larva > 0 && candidate.len)
```
clarity

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 5]
```diff
@@ -1168,3 +1159,39 @@ to_chat will check for valid clients itself already so no need to double check f
...
+		candidate.Cut(1,2)
+		xeno_job.occupy_job_positions(1)
+		stored_larva--
+		try_to_give_larva(next_in_line)
+	for(var/i in 1 to candidate.len)
+		candidate[i].larva_position = i
+		
+///Attempt to give a larva to the next in line, if not possible, free the xeno position and propose it to another candidate
+/datum/hive_status/normal/proc/try_to_give_larva(mob/next_in_line)
+	set waitfor = FALSE
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:1193]: Bad

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 6]
```diff
@@ -1168,3 +1159,39 @@ to_chat will check for valid clients itself already so no need to double check f
...
+	if(!attempt_to_spawn_larva(next_in_line, TRUE))
+		var/datum/job/xeno_job = SSjob.GetJobType(/datum/job/xenomorph)
+		xeno_job.free_job_positions(1)
+		give_larva_to_next_in_queue()
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:1197]: Early return

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 7]
```diff
@@ -19,6 +19,8 @@
 	var/list/list/xenos_by_zlevel
 	var/tier3_xeno_limit
 	var/tier2_xeno_limit
+	///Queue of all observer wanting to join xeno side
+	var/list/mob/dead/observer/candidate = list()
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:23]: Why isnt this a lazylist

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/hive_datum.dm [hunk 8]
```diff
@@ -1168,3 +1160,40 @@ to_chat will check for valid clients itself already so no need to double check f
...
+	var/stored_larva = xeno_job.total_positions - xeno_job.current_positions
+	var/mob/dead/observer/observer_in_queue
+	while(stored_larva > 0 && LAZYLEN(candidate))
+		observer_in_queue = LAZYACCESS(candidate, 1)
+		LAZYREMOVE(candidate, observer_in_queue)
+		xeno_job.occupy_job_positions(1)
+		stored_larva--
+		INVOKE_ASYNC(src, .proc/try_to_give_larva, observer_in_queue)
+	for(var/i in 1 to LAZYLEN(candidate))
+		observer_in_queue =LAZYACCESS(candidate, i)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/hive_datum.dm:1190]:
```suggestion
		observer_in_queue = LAZYACCESS(candidate, i)
```

**Comment by Swagile** [comment]
hello?

based department?

**Comment by Smudged** [comment]
Ah, waiting in the Pool DMV, good times. 

Good change.

**Comment by davidchan13** [comment]
I definitely like this.

Is there anyway to weight this so every death as a xeno in the round sets you as a lower priority? 
I.E. someone who observed or died as marine wants to play xeno, they put higher on the list than say someone who has died 3+ times already in the same round?

**Comment by BraveMole** [comment]
> I definitely like this.
> 
> Is there anyway to weight this so every death as a xeno in the round sets you as a lower priority?
> I.E. someone who observed or died as marine wants to play xeno, they put higher on the list than say someone who has died 3+ times already in the same round?

Looks like a good idea. Definitly not in the scope of this PR, but i might do it later

**Review by TiviPlus — APPROVED** [review]

**Comment by MiniMeatwad** [comment]
Joined queue with 0 burrowed larva and no pools, position 1. Later, 1 burrowed larva appears, and I try to join again, says "You are already in queue!". Someone else gets the larva from the hive leader calling burrowed. I try again and it puts me in position 2. I also prefer the old system, since you can control when you become a xenomorph and wont randomly spawn on a silo under attack unexpectedly, and you have to be immediately available in order to become a xenomorph.

**Comment by BraveMole** [comment]
> Joined queue with 0 burrowed larva and no pools, position 1. Later, 1 burrowed larva appears, and I try to join again, says "You are already in queue!". Someone else gets the larva from the hive leader calling burrowed. I try again and it puts me in position 2. I also prefer the old system, since you can control when you become a xenomorph and wont randomly spawn on a silo under attack unexpectedly, and you have to be immediately available in order to become a xenomorph.

Yeah there's a lot of broken thing in the system right now

**Comment by MiniMeatwad** [comment]
Nevermind, thought that was how it worked since I saw a larva spawn afk and die.

**Pull request merged**